src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java

```diff
@@ -0,0 +1,132 @@
+public class StringInterpolationProviderImpl
+    implements StringInterpolationProvider
+{
+    private static final String TYPE_ENV = "env";
+
+    private static final String TYPE_PROP = "prop";
+
+    private static final String DIRECTIVE_DEFAULT = "default";
+
+    /** Logger. */
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private Map<String, String> placeholderEntries = new HashMap<>();
+    private BundleContext context;
+
+    // ---------- SCR Integration ---------------------------------------------
+
+    /**
+     * Activates this component (called by SCR before)
+     */
+    @Activate
+    protected void activate(final BundleContext bundleContext, final StringInterpolationProviderConfiguration config) {
+        this.context = bundleContext;
+
+        String[] valueMap = config.placeHolderKeyValuePairs();
+        // Clear out any existing values
+        placeholderEntries.clear();
+        for(String line: valueMap) {
+            // Ignore no or empty lines
+            if(line != null && !line.isEmpty()) {
+                // Ignore comments
+                if(line.charAt(0) != '#') {
+                    int index = line.indexOf('=');
+                    if (index <= 0) {
+                        logger.warn("Placeholder Entry does not contain a key: '{}' -> ignored", line);
+                    } else if (index > line.length() - 2) {
+                        logger.warn("Placeholder Entry does not contain a value: '{}' -> ignored", line);
+                    } else {
+                        placeholderEntries.put(line.substring(0, index), line.substring(index + 1));
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Modifies this component (called by SCR to update this component)
+     */
+    @Modified
+    protected void modified(final BundleContext bundleContext, final StringInterpolationProviderConfiguration config) {
+        this.activate(bundleContext, config);
+    }
+
+    /**
+     * Deactivates this component (called by SCR to take out of service)
+     */
+    @Deactivate
+    protected void deactivate(final BundleContext bundleContext) {
+        this.context = null;
+        this.placeholderEntries = new HashMap<>();
+    }
+
+    /**
+     * This is the method that is used by the Map Entries service to substitute values with
+     * the proper format
+     * @param text Text to be converted
+     * @return Should be either the substituted text or the original given text
+     */
+    @Override
+    public String substitute(String text) {
+        logger.trace("Substitute: '{}'", text);
+        Object result = Interpolator.replace(text, (type, name, dir) -> {
+            String v = null;
+            if (TYPE_ENV.equals(type)) {
+                v = getVariableFromEnvironment(name);
+            } else if (TYPE_PROP.equals(type)) {
+                v = getVariableFromProperty(name);
+            } else {
```

Comment:
we should define a type for this, like "config:" and not let all types fallback to this. The fallback prevents adding new types later on (if required)

Reply:
I did not like the fallback either but I wanted to avoid forcing the user to write 'config:' for each value. That said etc-mapping is not easy by any chance and so I guess we add that.

Reply:
Done. This required the tests to be adjusted as now $[config: is required.